docs/api/vulnerability-info.md

```diff
@@ -0,0 +1,114 @@
+---
+title: Vulnerability Info, NuGet API
+description: The data that allows clients to check packages for known vulnerabilities.
+author: zivkan
+ms.author: zivkan
+ms.date: 06/16/2023
+ms.topic: reference
+---
+
+# Vulnerability information
+
+NuGet client, starting from version 6.7, can download known package vulnerability information to use in scenarios such as checking packages during restore operations.
+
+The API consists of at least two files, the [vulnerability index](#vulnerability-index) and one or more [vulnerability data files](#vulnerability-data).
+Known vulnerability data can be partitioned into multiple files, and the vulnerability index provides clients with information needed to cache files, and update the cache, efficiently.
+
+The resource used for building this URL is the `VulnerabilityInfo` resource found in the [service index](service-index.md).
+
+## Versioning
+
+The following `@type` values are used:
+
+@type value                       | Notes
+--------------------------------- | -----
+VulnerabilityInfo/6.7.0 | The initial release
+
+## Suggested partitioning strategy
+
+The files listed in the vulnerability index should ideally be optimized to maximize caching, and therefore minimize updates to large files.
+This will allow clients to minimize the frequency it needs to download updates.
+
+A suggested strategy for vulnerability data partitioning is to have two files, `base.json` and `updates.json`.
+The `base.json` file is updated periodically (for example once a month), and contains all known vulnerabilities at the time the file is regenerated.
+The `updates.json` file should any new advisories published since `base.json` was last regenerated.
+This will allow clients to download the large `base.json` infrequently, while the frequently changing `updates.json` file is always relatively small.
+
+NuGet client does not provide a way for one file to update or remove entries from another file, so if any modifications are needed to any advisory in `base.json`, then `base.json` can be regenerated and all entries in `updates.json` merged.
+
+## Vulnerability index
+
+The vulnerability index is a JSON array of objects with the following properties:
```

Comment:
I am sorry that I haven't checked what the final design is. My understanding is that one day we will have to purge very old vulnerability data as the file size will grow forever. Will it be better that we have a property to say what the time range is for vulnerability data we provide? Otherwise, customers may complain that we miss a vulnerable package as it was marked as vulnerable long time ago and we just assume that no one is using that.

Reply:
Currently new vulnerabilities in the NuGet ecosystem are growing slowly, so I hope that improvements in internet connectivity, HTTP compression algorithms, CPU speed (or at least cores, so we can do more work in parallel), and disk speed&capacity will outpace the growth of the file due to new vulnerabilities. So, I'm not sure that we will need to purge very old vulnerabilities. I think we'll be able to maintain perf for years to come, but it does depend on how quickly the file size ends up growing.

But if we want to, we can change the protocol design to cater for it. NuGetAudit is currently only available in preview versions of NuGet, but we have a very, very short time to make changes before NuGet 6.7 is locked in, at which time the protocol design will be used by GA versions of NuGet.

Another change which I wish we added initially, is to wrap all arrays (and the data file root object) in an object with a count property. This would let clients pre-allocate arrays of the correct size, reducing the number of resize events, and therefore improve performance. Something to keep in mind if we ever make other breaking changes (and therefore a new resource version), but when I do performance tracing with the current file(s), I don't see this being significant in the context of the overall restore performance.

Reply:
We can enrich that "comment" property with a better visibility for customers, if we have to do so one day. No protocol needs to be broken.

Reply:
@zhhyu, I think it would be surprising to client implementors based on this current doc to ever have the comment overloaded to be used for anything more than an informational string. I agree with @zivkan's point that we can just release a new resource version (e.g. `VulnerabilityInfo/8.0.0`) with a radically different shape if needed. Or some tactical incremental change to the current protocol that may or may not be backwards compatible. The resource version concept allows us to be very creative in the future without much worry.